torch/distributed/fsdp/_optim_utils.py

```diff
@@ -858,31 +858,26 @@ def _broadcast_unsharded_pos_dim_tensor_state(
 def _rekey_sharded_optim_state_dict(
     sharded_osd: Dict[str, Any],
     model: torch.nn.Module,
-    optim_input: Optional[
-        Union[
-            List[Dict[str, Any]],
-            Iterable[torch.nn.Parameter],
-        ]
-    ] = None,
+    optim: torch.optim.Optimizer,
```

Comment:
one suggestion for backwards compatibility would be to maintain the old codepath, add this new argument as optional, and log a warning indicating to use this argument and mention that the old one will be deprecated.

Although, since FSDP is beta and doesn't guarantee BC yet, this change + fixing the callsites we know of should be acceptable as well.

Reply:
This is an internal API, it's ok to change it at this point.

Reply:
I followed @rohan-varma and @fegin's suggestion to preserve the existing code path so that we do not need to migrate internal customers in the short term. We are playing to revamp the API soon anyway, so to move faster, it makes sense to preserve the `optim_input` code path for the customers using it. The goal after this PR is to accommodate users that want to use the existing APIs (before we revamp) but do not want to pass `optim_input`.